File: session-06-capstone/go/resilience/rate_limiter.go
package resilience

import (
	"fmt"
	"sync"
	"time"

	"coffeeshop-capstone/kitchen"
)

// Exercise: Rate limiter wrapper. Rejects after MaxRequests in WindowDuration.
type RateLimiterKitchenService struct {
	Inner          kitchen.KitchenService
	MaxRequests    int
	WindowDuration time.Duration

	mu           sync.Mutex
	requestCount int
	windowStart  time.Time
}

func (s *RateLimiterKitchenService) Prepare(orderID int, drink string) (*kitchen.Preparation, error) {
	s.mu.Lock()
	now := time.Now()
	if s.windowStart.IsZero() || now.Sub(s.windowStart) > s.WindowDuration {
		s.windowStart = now
		s.requestCount = 0
	}
	s.requestCount++
	count := s.requestCount
	s.mu.Unlock()

	if count > s.MaxRequests {
		return nil, fmt.Errorf("rate limit exceeded: max %d requests per %s", s.MaxRequests, s.WindowDuration)
	}

	fmt.Printf("[RateLimiter] Request %d/%d\n", count, s.MaxRequests)
	return s.Inner.Prepare(orderID, drink)
}

File: session-06-capstone/python/resilience/rate_limiter.py
import time

from kitchen.service import KitchenService, Preparation


class RateLimiterKitchenService(KitchenService):
    """Exercise: Rate limiter wrapper. Rejects if more than max_requests in window_seconds."""

    def __init__(self, inner: KitchenService, max_requests: int = 10, window_seconds: float = 60.0):
        self.inner = inner
        self.max_requests = max_requests
        self.window_seconds = window_seconds
        self._request_count = 0
        self._window_start = time.time()

    def prepare(self, order_id: int, drink: str) -> Preparation:
        now = time.time()
        if now - self._window_start > self.window_seconds:
            self._window_start = now
            self._request_count = 0

        self._request_count += 1
        if self._request_count > self.max_requests:
            raise RuntimeError(
                f"Rate limit exceeded: max {self.max_requests} requests per {self.window_seconds}s"
            )

        print(f"[RateLimiter] Request {self._request_count}/{self.max_requests}")
        return self.inner.prepare(order_id, drink)
